### PR TITLE
RequestorStep.js: class to functional component refactor

### PR DESCRIPTION
**Pull request by kacper-mikolajczak**
### Details
PR implements refactor from class to functional component of `RequestorStep.js`

### Fixed Issues

$ https://github.com/Expensify/App/issues/16247
PROPOSAL: https://github.com/Expensify/App/issues/16247#issue-1631654397

### Tests

- [x] Verify that no errors appear in the JS console

1. Turn on `Use Staging Server` in `Preferences` in `Settings`
2. Open workspace in `Settings` -> `Workspaces`
3. Open `Bank account`
4. Fill out steps in manual mode until `Personal Info` appears
5. Check if inputs work correctly
6. Check if validation works
7. Check if submission works

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/62747088/5d87fd3a-f688-4b15-89bf-8af9e143f3e9

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/62747088/7a8fd3db-3c18-44c4-ac2b-98abc45e1f1a

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/62747088/5f284830-4fc8-43af-ae6f-359515ca9968

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/62747088/3dc160a9-8647-4841-b079-3be35d566c83

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/62747088/37077a20-942a-4d25-9b51-4a270fbc5859

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:93]: Can we re-use `INPUT_KEYS`?
```suggestion
            firstName: getDefaultStateForField(INPUT_KEYS.firstName),
            lastName: getDefaultStateForField(INPUT_KEYS.lastName),
            street: getDefaultStateForField(INPUT_KEYS.street),
            ....
```

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:101]:
```suggestion
                bankAccountID: lodashGet(reimbursementAccount, 'achData.bankAccountID', 0),
```

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:151]: Can we add it to `CONST.ts`? Moreover, there are other links there.

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:158]: Can we add it to `CONST.ts`? Moreover, there are other links there.

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:153]:
```suggestion
                        {translate('requestorStep.learnMore')}
```

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:160]:
```suggestion
                        {translate('requestorStep.isMyDataSafe')}
```

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:110]:
```suggestion
    const renderLabelComponent = useCallback(
```

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:173]:
```suggestion
                    LabelComponent={renderLabelComponent}
```

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [inline src/pages/ReimbursementAccount/RequestorStep.js:206]: `displayName` is required for functional components.

**Review by rezkiy37 — COMMENTED** [review]
Left comments.

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [inline src/pages/ReimbursementAccount/RequestorStep.js:119]: Add the `{}` to the if, because it is in multiple lines, it hard to read

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [inline src/pages/ReimbursementAccount/RequestorStep.js:205]: please add the `displayName`
and remove the `withLocalize` and use the hook `useLocalize`

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [inline src/pages/ReimbursementAccount/RequestorStep.js:41]: I don't see the value of adding the styles into an object, also isn't the standard in the codebase. I will keep the styles directly into the components

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [inline src/pages/ReimbursementAccount/RequestorStep.js:205]: This is spreading to other components @mountiny, what format should we stick for the `forwardRef`? In our codebase we use:
```
export default React.forwardRef((props, ref) => (
    <RequestorStep
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
        forwardedRef={ref}
    />
));
```
But I wanna bring into a discussion this format proposed by @kacper-mikolajczak

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [reply src/pages/ReimbursementAccount/RequestorStep.js:110]: Isn't a complex component or anything, I would not wrap it into a `useCallback`

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (mountiny) [reply src/pages/ReimbursementAccount/RequestorStep.js:205]: I think these are best in Slack where more people can weigh in

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js:206]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js:173]: Hi @rezkiy37, is it a convention inside a project to name such components with `render` prefix? Thanks for all suggestions!

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js:209]: Done ✅ @gedu I needed to modify `PropTypes` could you check if such modification is not breaking any convention? Thanks

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js:102]: I've wrapped in `useCallback` with an idea to keep its reference stable for usage inside other component, in this case `CheckboxWithLabel`. I can drop it tho :)

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js:151]: That's a great idea! Is there any particular place inside CONST.ts that we should place the links?

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [reply src/pages/ReimbursementAccount/RequestorStep.js:102]: `CheckboxWithLabel` doesn't use it as a hook, I think we can drop it

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [inline src/pages/ReimbursementAccount/RequestorStep.js:24]: I think you can just do
```
const propTypes = {
      ...StepPropTypes,
      [...]
      }
```

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [reply src/pages/ReimbursementAccount/RequestorStep.js:151]: @kacper-mikolajczak, I see a whole block with URLs, you can put your right after those.
Screenshot 2023-09-15 at 11 08 51

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (rezkiy37) [reply src/pages/ReimbursementAccount/RequestorStep.js:166]: Basically, it is a common practise across React/React Native projects to call such render functions like `renderSomething`. Within Expensify code you can see `renderItem`, `renderListEmptyComponent`, `renderSuggestionMenuItem` and so on.

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js:24]: `StepPropsTypes` holds required typings for `withLocalize` that we need to omit here, that's why I've created this messy syntax :D

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (kacper-mikolajczak) [reply src/pages/ReimbursementAccount/RequestorStep.js]: Done ✅

File: src/pages/ReimbursementAccount/RequestorStep.js
Comment (gedu) [reply src/pages/ReimbursementAccount/RequestorStep.js:193]: [Thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1695110739750909) created to discuss this

**Review by Ollyws — APPROVED** [review]
LGTM

**Comment by mountiny** [comment]
@chiragsalian is assigned for a review here so I am going to leave that to him

**Comment by kacper-mikolajczak** [comment]
Hi @chiragsalian, could you please take a look? Thanks!

**Comment by chiragsalian** [comment]
Sorry i didn't get a chance to review this today. It's on my plate for tomorrow.

**Review by chiragsalian — APPROVED** [review]
Code LGTM and works well for me.
I see there are other reviewers tagged here. @gedu, @mountiny and @rezkiy37, did either of you guys also want to review the PR before we merge it?

**Comment by gedu** [comment]
@chiragsalian  LGTM

**Comment by mountiny** [comment]
Just going to merge this, thanks for the reviews!

**Pull request merged**
